Add shared routes, nav items and views to dev console

rohitkrai03 opened the pull request:
Related Jira story https://jira.coreos.com/browse/ODC-128

- Adds shared routes based on the active perspective.
- Updates the utility function to generate Resource links.
- Updates all the other links on any shared view pages using a utility function which prepends given path with active perspective.
- Updates the namespace dropdown bar to work with other perspectives and dispatch correct actions when changing a namespace.
- Add menu items for shared views in dev console navigation based on latest updates from UX team.

<img width="200" alt="Screenshot 2019-04-05 at 2 04 13 AM" src="https://user-images.githubusercontent.com/6041994/55586662-2796ce80-5747-11e9-924f-045e6de11e83.png">

rohitkrai03 committed:
Add shared routes, nav items and views to dev console

rohitkrai03 committed:
Created PerspectiveLink component to connect to store and prepend link with active perspective

christianvogt commented on frontend/public/components/nav.jsx:
```diff
@@ -93,15 +94,15 @@ NavLink.propTypes = {
   get to() {
     const { resource, activeNamespace } = this.props;
-    return formatNamespacedRouteForResource(resource, activeNamespace);
+    return pathWithPerspective(formatNamespacedRouteForResource(resource, activeNamespace));
```
Why don't you use `PerspectiveLink in `NavLink#render`?

christianvogt commented on frontend/public/components/nav.jsx:
```diff
@@ -113,13 +114,13 @@ ResourceNSLink.propTypes = {
 
   get to() {
-    return `/k8s/cluster/${this.props.resource}`;
+    return pathWithPerspective(`/k8s/cluster/${this.props.resource}`);
```
Why don't you use `PerspectiveLink in `NavLink#render`?

christianvogt commented on frontend/public/components/utils/link.jsx:
```diff
@@ -16,14 +16,33 @@ import { ALL_NAMESPACES_KEY } from '../../const';
+    return [...acc, ...defaultNamespacedPrefixes.map(prefix => `/${perspective}${prefix}`)];
+  }
+  return [...acc, ...defaultNamespacedPrefixes];
+}, []);
```
Does this affect `ui-actions#formatNamespaceRoute`. Looks like it might as that code splits on `/` and references '[0]`.

rohitkrai03 replied:
No it doesn't affect `ui-actions#formatNamespaceRoute`,  if we don't have Namespace Prefixes with perspectives then the formatting doesn't work since it doesn't recognises the prefix. 
So, the logic of `ui-actions#formatNamespaceRoute` first filters all `links#namespacedPrefixes` to get the current locations prefix. Then it gets `activeNamespace` and checks if previous namespace are not same and then generates the final route for namespace. So, having all the namespace prefixes with perspective enables us to be in the same perspective even when changing namespace.

christianvogt commented on frontend/public/components/utils/perspective.tsx:
```diff
@@ -0,0 +1,7 @@
+export const pathWithPerspective = path => {
+  const activePerspective = getActivePerspective(store.getState());
+  return activePerspective !== 'admin' ? `/${activePerspective}${path}` : path;
+};
```
Please add a `FIXME` comment related to using global store.

christianvogt commented on frontend/public/extend/devconsole/shared/components/PerspectiveLink.tsx:
```diff
@@ -0,0 +1,31 @@
+  activePerspective: string;
+}
+
+type PerspectiveLinkProps = StateProps & LinkProps;
```
`export` this type

christianvogt commented on frontend/public/extend/devconsole/shared/components/PerspectiveLink.tsx:
```diff
@@ -0,0 +1,31 @@
+) => {
+  const _pathWithPerspective = (activePerspective, path) => {
+    return activePerspective !== 'admin' ? `/${activePerspective}${path}` : path;
+  };
```
This should be a static function and not in the render function. Or just move the logic inline of the return value.

christianvogt commented on frontend/public/extend/devconsole/components/DevConsoleNav.tsx:
```diff
@@ -22,41 +24,52 @@ export const PageNav = (props: DevConsoleNavigationProps) => {
+        </DevNavSection>
+        <DevNavSection title="Pipelines">
+          <HrefLink href="/dev/pipelines" name="Piplines" activePath="/dev/pipelines/" />
+          <HrefLink href="/dev/pipelines/run" name="Pipline Runs" activePath="/dev/pipelines/run" />
```
Remove this entry.

Only `Pipelines` should be present and it should not be within a section. It should be at the top level.

christianvogt commented on frontend/public/extend/devconsole/components/DevConsoleNav.tsx:
```diff
@@ -22,41 +24,52 @@ export const PageNav = (props: DevConsoleNavigationProps) => {
           />
         </DevNavSection>
+        <DevNavSection title="Builds">
+          <ResourceNSLink resource="buildconfigs" name={BuildConfigModel.labelPlural} />
```
Name this `Builds`? Ya it's weird. I last heard, build config was what we need to show. We need to confirm with Serena thought before committing to this change as I think it'll be confusing for those switching from Admin perspective.

It should also be a top level item and not within a section now that there's only one.

christianvogt commented on frontend/public/extend/devconsole/components/DevConsoleNav.tsx:
```diff
@@ -22,41 +24,52 @@ export const PageNav = (props: DevConsoleNavigationProps) => {
+        <DevNavSection title="Builds">
+          <ResourceNSLink resource="buildconfigs" name={BuildConfigModel.labelPlural} />
+          <ResourceNSLink resource="builds" name={BuildModel.labelPlural} />
+          <ResourceNSLink resource="imagestreams" name={ImageStreamModel.labelPlural} startsWith={['imagestreams', 'imagestreamtags']} />
```
Remove `builds` and `imagestreams`

christianvogt commented on frontend/public/extend/devconsole/components/DevConsoleNav.tsx:
```diff
@@ -22,41 +24,52 @@ export const PageNav = (props: DevConsoleNavigationProps) => {
-            href="/dev/submenu_1"
-            name="Sub Menu 1"
-            activePath="/dev/submenu_1/"
+        <DevNavSection title="Catalog">
```
Remove the entire `Catalog` section.

christianvogt commented on frontend/public/components/utils/resource-link.jsx:
```diff
@@ -33,7 +34,7 @@ export const resourcePathFromModel = (model, name, namespace) => {
   }
 
-  return url;
+  return pathWithPerspective(url);
```
Please add a `FIXME` comment related to using global store.

christianvogt commented on frontend/public/components/nav.jsx:
```diff
@@ -113,13 +114,13 @@ ResourceNSLink.propTypes = {
+export class ResourceClusterLink extends NavLink {
   static isActive(props, resourcePath) {
-    return resourcePath === props.resource || _.startsWith(resourcePath, `${props.resource}/`) || matchesModel(resourcePath, props.model);
+    return resourcePath === props.resource || _.startsWith(resourcePath, `${props.resource}/`) || matchesModel(resourcePath, props.model) || resourcePath.includes(props.resource);
```
Isn't this risky? shouldn't we strip the legal perspectives and do the comparisons against the path without the perspective segment?

rohitkrai03 committed:
Use PerspectiveLink in NavLink, fix nav links isActive and remove unwanted menu items

christianvogt commented on frontend/public/components/nav.jsx:
```diff
@@ -31,12 +30,18 @@ export const matchesPath = (resourcePath, prefix) => resourcePath === prefix ||
 
+const stripPS = href => {
+  href = stripBasePath(href);
+  return href.replace(/^admin|dev\//, '');
```
Isn't there a list of perspectives you can make use of for this purpose? Furthermore admin isn't a valid perspective segment.

rohitkrai03 committed:
Fix PerspectiveLink warnings, use legal perspective name to generate stripping pattern

rohitkrai03 committed:
Fix Lint and rebase errors

rohitkrai03 committed:
Fix bug in MegaMenuItem related to actions not being dispatched

rohitkrai03 committed:
Using PerspectiveLink in actions dropdown to maintain perspective when choosing an action from namespace bar.
Changed name of pathWithPerspective -> pathWithPerspectiveFromStore
Created pathWithPerspective and using it to contain the logic in one

rohitkrai03 commented:
@christianvogt 
- I renamed `pathWithPerspective` -> `pathWithPerspectiveFromStore`.
- Created a new function `pathWithPerspective` which takes `activePerspective` and `path` encapsulating the logic of default perspective. Using it everywhere so that it would be easier to change the logic later.
- Fixed a bug in `MegaMenuItem` after the rebase.
- Also, added logic to generate links in `Add` action dropdown on the top. So, now any action stays in the same perspective. 

The PR is ready for review again.

rohitkrai03 committed:
Using PerspectiveLink in clusterserviceversion.tsx

rohitkrai03 committed:
Using PerspectiveLink for more links in shared views

rohitkrai03 committed:
Fix broken test
